Remove Jetty as a runtime dependency of Javalin

mvysny opened the pull request:
Make Javalin run without Jetty, to be able to be embedded into other servlet containers. Fixes https://github.com/tipsy/javalin/issues/402 .

tipsy commented on src/main/java/io/javalin/Javalin.java:
```diff
@@ -51,8 +51,8 @@
-    private Server jettyServer = JettyServerUtil.defaultServer();
-    private SessionHandler jettySessionHandler = JettyServerUtil.defaultSessionHandler();
+    private Server jettyServer = null;
+    private SessionHandler jettySessionHandler = null;
```
Add comments explaining this?

mvysny replied:
The `defaultServer()` function will fail if Jetty is not on the classpath. The same with default session handler. Therefore I'm deferring the call to those only after somebody attempts to `javalin.start()`

tipsy replied:
I know, but that's not obvious for new contributors reading the code. If we choose this approach we need to explain what's going on.

mvysny replied:
You're right. Yet, Passing `jettyServer` and `jettySessionHandler` as constructor parameters seem to work pretty nicely, so I was able to remove the now-useless null check.

tipsy commented on src/main/java/io/javalin/Javalin.java:
```diff
@@ -123,21 +123,13 @@ public Javalin start() {
+                if (jettyServer == null) {
+                    jettyServer = JettyServerUtil.defaultServer();
+                }
+                JavalinServlet javalinServlet = createServlet(new JettyResourceHandler(staticFileConfig, jettyServer, ignoreTrailingSlashes));
```
Could you move this under the other null-check?

mvysny replied:
Passing `jettyServer` and `jettySessionHandler` as constructor parameters seem to work pretty nicely, so I was able to remove the now-useless null check.

tipsy commented on src/main/java/io/javalin/Javalin.java:
```diff
@@ -166,7 +177,9 @@ public Javalin stop() {
         log.info("Stopping Javalin ...");
         try {
-            jettyServer.stop();
+            if (jettyServer != null) {
```
I think a nullpointer here would be better than nothing.

mvysny replied:
stop() is generally expected to do nothing if the server is not started; throwing npe is unexpected.

tipsy replied:
Then `if (jettyServer == null) throw new AppropriateException()`. If we can extend Javalin and override start/stop the problem solves itself.

mvysny replied:
Passing `jettyServer` and `jettySessionHandler` as constructor parameters seem to work pretty nicely, so I was able to remove the now-useless null check.

tipsy commented:
Looking at this, I'm wondering if it would be better to extend `Javalin`. You could probably avoid these checks that way. Might need to make the Jetty stuff protected rather than private.

tipsy commented:
With this approach you could also override irrelevant methods:

```java
public class ServletExposingJavalin extends Javalin {

    public JavalinServlet createServlet() {
        return new JavalinServlet(
            this,
            pathMatcher,
            exceptionMapper,
            errorMapper,
            debugLogging,
            requestLogger,
            dynamicGzipEnabled,
            autogeneratedEtagsEnabled,
            defaultContentType,
            maxRequestCacheBodySize,
            prefer405over404,
            singlePageHandler,
            null
        );
    }

    @Override
    public Javalin enableStaticFiles(@NotNull String classpathPath) {
        throw new RuntimeException("Cannot use static files in standalone mode");
    }

    @Override
    public Javalin contextPath(@NotNull String contextPath) {
        throw new RuntimeException("Cannot use context-path standalone mode");
    }

    @Override
    public Javalin start() {
        throw new RuntimeException("Cannot start server in standalone mode");
    }

}
```

🤔

tipsy commented:
Can be made a lot less verbose if you use Kotlin: 

```kotlin
class ServletExposingJavalin : Javalin() {

    fun createServlet() = JavalinServlet(
        this,
        pathMatcher,
        exceptionMapper,
        errorMapper,
        debugLogging,
        requestLogger,
        dynamicGzipEnabled,
        autogeneratedEtagsEnabled,
        defaultContentType,
        maxRequestCacheBodySize,
        prefer405over404,
        singlePageHandler,
        null
    )

    override fun enableStaticFiles(classpathPath: String) = notAvailable("enableStaticFiles()")
    override fun contextPath(contextPath: String) = notAvailable("contextPath()")
    override fun start() = notAvailable("start()")
    private fun notAvailable(action: String): Nothing = throw RuntimeException("$action is not available in standalone mode")

}
```

I'm not sure if Java will try loading the classes of the parent though, could you test?

tipsy commented:
> I believe that extending class shouldn't remove a functionality from its superclass. Ideally Javalin's code should be moved to e.g. JavalinBase; but then we need generics to make all JavalinBase methods eventually return Javalin. So I guess your approach is simpler.

It's better than having methods that silently fail, IMO. I know it's not clean, I'm just trying to make the smallest change we can get away with in order to enable this feature. We'll have to rewrite it for 3.0, so I'm trying to keep things simple now.

> I propose to name the class EmbeddedJavalin instead of ServletExposingJavalin.

This won't be a widely used feature, so I don't have strong feelings on the name. I'm fine with `EmbeddedJavalin` or `EmbeddableJavalin`, or even things like `JavalinServletBuilder`.

tipsy commented:
@mvysny Do you think you can finish this before the weekend? I'm considering releasing 2.4.0 on Saturday or Sunday.

mvysny commented:
@tipsy I think the current state of the PR should be according to your requirements. I can't remove the null assignment to the jetty server unfortunately, since filling it up with an actual Jetty server instance would make it crash with NoClassDefFoundError outright when running from Tomcat.

Please let me know if there is anything else I can improve.

tipsy commented:
@mvysny the methods I used in my code snippets were just examples, there are a lot more of them. Almost everything from https://javalin.io/documentation#configuration needs to be overridden. I really want to avoid the null values in `Javalin`, let me play around with it a bit locally one of the next few days.

tipsy commented:
I've been unexpectedly busy this week, but I'll try to have a look in the weekend.

tipsy commented:
I was wondering if we could create a factory method `EmbeddedJavalin.create()` which cleared the jetty-fields of `Javalin` using reflection, before creating an instance. I couldn't quite get there though.

mvysny commented:
Sorry, I've been quite busy myself too. But I got this idea that we could pass `jettyServer` and `jettySessionHandler` as constructor parameters! And indeed it seem to work pretty nicely, so I was able to remove the now-useless null check. What do you think?

tipsy commented:
Great, this feels a lot cleaner than the previous solution!

tipsy commented:
The (un)helpful message override should be moved into the constructor of EmbeddedJavalin, users shouldn't have to do that themselves.

mvysny commented:
You're right - please review :)

tipsy commented:
I've been playing around a bit with this. My suggested overrides are pretty useless since any method on `Javalin` will return the a `Javalin` object, which doesn't have these overrides. Brain fart.

What we can do instead of build on `ensureActionIsPerformedBeforeServerStart` and create something like `ensureHasJettyServer` ? We could also just skip this, and let users figure this out on their own, as it's a pretty niche use case.. In which case you can just remove the overrides and I'll merge the PR.

mvysny commented:
Hi, actually in my case overrides work:
```
    val javalin = EmbeddedJavalin()
            .get("/rest") { ctx -> ctx.result("Hello!") }
            .start()
            .createServlet()
```
This code correctly fails with:
```
 java.lang.RuntimeException: start() is not available in standalone mode
	at io.javalin.EmbeddedJavalin.notAvailable(EmbeddedJavalin.kt:53)
	at io.javalin.EmbeddedJavalin.start(EmbeddedJavalin.kt:52)
```
The reason is that `get()` returns `this` which is `EmbeddedJavalin`.

With that being said, I would advise against investing a lot of time into this. The reason is that this is just a hack (if done properly it should be the other way around, there should be `JettyJavalin` adding support for Jetty). Since you're going to refactor this anyway in 3.0.0, maybe we could keep it as is (or maybe add a few more overrides - I can do that yet please list the methods to override).

The reason why I'm pushing for this: I'm planning to ditch RESTEasy in [Vaadin-on-Kotlin](http://vaadinonkotlin.eu) and replace it with Javalin; VoK apps are primarily WARs which run in servlet environment.

tipsy commented:
Do you also get the depreciations? I couldn't see them when checking out your code locally, the type was inferred to just `Javalin`.

tipsy commented:
I added these at least:

```
@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun contextPath(contextPath: String) = notAvailable("contextPath()")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun enableStaticFiles(classpathPath: String) = notAvailable("enableStaticFiles()")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun enableWebJars() = notAvailable("enableWebJars()")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun port() = notAvailable("port()")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun port(port: Int) = notAvailable("port(port)")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun server(server: Supplier<Server>) = notAvailable("server()")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun sessionHandler(sessionHandler: Supplier<SessionHandler>) = notAvailable("sessionHandler()")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun start() = notAvailable("start()")

@Deprecated("Not available in standalone mode", level = DeprecationLevel.ERROR)
override fun stop() = notAvailable("stop()")
```

mvysny commented:
Ah, yes, the `@Deprecated` annotations are useless, we can remove them. Let me create a PR

mvysny commented:
Done - @tipsy can you please review?

tipsy commented:
Added a few changes in https://github.com/tipsy/javalin/commit/848169cb3d03e128f4cc85efadb960eb6334e8a2, mostly formatting.

mvysny commented:
Awesome, thank you very much :+1:

tipsy commented:
Yes, it will

tipsy commented:
@mvysny This change is out as part of 2.4.0 now: https://javalin.io/news/2018/11/04/javalin-2.4.0-released.html

mvysny commented:
@tipsy Awesome, thank you for letting me know :+1:

tipsy commented:
It's an interesting feature, thanks for working on it!

tipsy commented:
@mvysny Have you taken this into use yet? Running into any problems? Would you be interested in writing a tutorial for https://javalin.io at some point?

mvysny commented:
@tipsy I have. In fact I've replaced RESTEasy with Javalin and it has been working really nicely. I've already updated the documentation in my project here: https://github.com/mvysny/vaadin-on-kotlin/tree/master/vok-rest . Regarding the tutorial - sure, what do you have in mind?

tipsy commented:
> @mvysny @tipsy I have. In fact I've replaced RESTEasy with Javalin and it has been working really nicely. I've already updated the documentation in my project here: https://github.com/mvysny/vaadin-on-kotlin/tree/master/vok-rest .

Happy to hear that, seems like a good use-case.

> Regarding the tutorial - sure, what do you have in mind?

Nothing fancy, just a simple guide showing how to pull in tomcat, exclude jetty, and set up the Servlet. Preferably with an example repo people can clone. There were a few people asking about this before you implemented the functionality, so I think it would be good to show it off in a tutorial.

mvysny commented:
Sure, that I can do. I can use this example repo: https://gitlab.com/mvysny/javalin-tomcat-embed-test ; I can write a simple guide no worries. Could you point me out to Javalin docs sources, so that I can create a PR for those please?

tipsy commented:
Great! Source is https://github.com/javalin/javalin.github.io.

mvysny commented:
Thanks @tipsy created a PR with the first draft: https://github.com/javalin/javalin.github.io/pull/20
